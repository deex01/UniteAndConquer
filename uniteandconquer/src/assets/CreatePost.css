.create-post-page {
  height: 100%;
  display: grid;
  grid-template-columns: 0.2fr 0.8fr;
}

.all-fields {
  text-align: center;
}

textarea {
  border: none;
  background-color: #ffd9a0;
  margin: auto;
  width: 90%;
  height: 30px;
  padding: 10px;
  font-size: 20px;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  resize: none;
}

.desc {
  height: 75px;
  width: 46.5%;
}

.create-post {
  text-align: center;
}

.post-input {
  margin: auto;
}

.post-text-fields {
  display: grid;
  grid-template-columns: 0.5fr 0.5fr;
  margin: auto;
  width: 50%;
}

.post-field {
  margin: 15px;
}

input {
  border: none;
  background-color: #ffd9a0;
  margin: auto;
  width: 90%;
  height: 30px;
  padding: 10px;
  font-size: 20px;
}

.label {
  font-style: italic;
  font-size: 20px;
}

.desc-label {
  font-style: italic;
  font-size: 20px;
}

.tags-label {
  font-style: italic;
  font-size: 20px;
}

.description {
  width: 100%;
  text-align: center;
}

textarea:hover {
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
}

.create {
  border: none;
  background: #ffcb7d;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 150px;

  font-style: italic;
  font-weight: bold;
  font-size: 20px;
  margin: 5px;
}

.create:hover {
  box-shadow: 5px 5px 5px rgba(0, 0, 0, 0.25);
  background-color: grey;
  transition-duration: 0.4s;
}

.cancel {
  border: none;
  background: #b4aea7;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 150px;

  font-style: italic;
  font-weight: bold;
  font-size: 20px;
  margin: 5px;
}

.cancel:hover {
  box-shadow: 5px 5px 5px rgba(0, 0, 0, 0.25);
  background-color: grey;
  transition-duration: 0.4s;
}

.post-tags {
  display: grid;
  grid-template-columns: 0.25fr 0.25fr 0.25fr 0.25fr;
  margin: auto;
  width: 50%;
  margin-top: 10px;
}

.create-post-tags {
  margin: 10px;
}
