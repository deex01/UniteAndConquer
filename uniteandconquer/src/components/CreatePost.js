/* eslint-disable no-unused-vars */
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import Sidebar from './Sidebar';
import '../assets/App.css';
import '../assets/CreatePost.css';

const PostDB = require('../modules/PostDB');

function CreatePost() {
  const [item, setItem] = useState('');
  const [price, setPrice] = useState(0);
  const [quantity, setQuantity] = useState('');
  const [link, setLink] = useState('');
  const [description, setDescription] = useState('');
  const [tags, setTags] = useState([]);
  const navigate = useNavigate();
  const createPost = () => {
    if (!item || !price || !quantity || !link || !description || !tags) {
      throw new Error('You need to fill in all the blank');
    } else {
      PostDB.addPost(
        item,
        Number(quantity),
        Number(quantity),
        Number(price),
        link,
        description,
        tags,
        (success, id, err) => {
          if (success) {
            navigate(`/post-details${id}`);
          } else {
            console.log(err);
          }
        },
      );
    }
  };

  const cancel = () => {
    navigate('/');
  };

  return (
    <div className="create-post-page">
      {Sidebar()}
      <div className="create-post">
        <div className="post-input">
          <h3>Create a Post</h3>
          <div className="all-fields">
            {' '}
            <div className="post-text-fields">
              <div className="post-field">
                <div className="label">Item Name</div>
                <input data-testid="item-name-input" onChange={(e) => setItem(e.target.value)} />
              </div>
              <div className="post-field">
                <div className="label">Price/Item</div>
                <input data-testid="price-input" onChange={(e) => setPrice(e.target.value)} />
              </div>
              <div className="post-field">
                <div className="label">Target Quantity</div>
                <input data-testid="quantity-input" onChange={(e) => setQuantity(e.target.value)} />
              </div>
              <div className="post-field">
                <div className="label">Item Link</div>
                <input data-testid="link-input" onChange={(e) => setLink(e.target.value)} />
              </div>
            </div>
            <div className="description">
              <div className="desc-label">Description</div>
              <textarea
                data-testid="description-input"
                className="desc"
                onChange={(e) => setDescription(e.target.value)}
              />
            </div>
            <div className="create-post-tags">
              <div className="tags-label">Tags</div>
              <div className="post-tags">
                <div className="tag">Tag1</div>
                <div className="tag">Tag2</div>
                <div className="tag">Tag3</div>
                <div className="tag">Tag4</div>
                <div className="tag">Tag5</div>
                <div className="tag">Tag6</div>
              </div>
            </div>
          </div>
          <br />
          <div className="create-post-bottom">
            <button className="create" type="button" onClick={createPost}>
              Post
            </button>
            <button className="cancel" type="button" onClick={cancel}>
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  );
}

export default CreatePost;
