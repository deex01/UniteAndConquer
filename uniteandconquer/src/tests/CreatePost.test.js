/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import CreatePost from '../components/CreatePost';

// https://stackoverflow.com/questions/66284286/react-jest-mock-usenavigate

const mockedUsedLink = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: () => mockedUsedLink,
  useNavigate: () => mockedUsedNavigate,
}));

test('create post header', () => {
  render(<CreatePost />);
  const headerElement = screen.getByText('Create a Post');
  expect(headerElement).toBeInTheDocument();
});

// item name

test('input item name', () => {
  render(<CreatePost />);
  const input = screen.getByTestId('item-name-input');
  fireEvent.change(input, { target: { value: 'Lights' } });
  expect(input.value).toBe('Lights');
});

// price

test('input item price', () => {
  render(<CreatePost />);
  const input = screen.getByTestId('price-input');
  fireEvent.change(input, { target: { value: '34' } });
  expect(input.value).toBe('34');
});

// quantity

test('input item quantity', () => {
  render(<CreatePost />);
  const input = screen.getByTestId('quantity-input');
  fireEvent.change(input, { target: { value: '32' } });
  expect(input.value).toBe('32');
});

// link

test('input item link', () => {
  render(<CreatePost />);
  const input = screen.getByTestId('link-input');
  fireEvent.change(input, { target: { value: 'exampleurl.com/item' } });
  expect(input.value).toBe('exampleurl.com/item');
});

// description

test('input item description', () => {
  render(<CreatePost />);
  const input = screen.getByTestId('description-input');
  fireEvent.change(input, { target: { value: 'this item is great' } });
  expect(input.value).toBe('this item is great');
});

test('cancel post', () => {
  render(<CreatePost />);
  const name = screen.getByTestId('item-name-input');
  fireEvent.change(name, { target: { value: 'Lights' } });
  expect(name.value).toBe('Lights');
  const price = screen.getByTestId('price-input');
  fireEvent.change(price, { target: { value: '34' } });
  expect(price.value).toBe('34');
  const quantity = screen.getByTestId('quantity-input');
  fireEvent.change(quantity, { target: { value: '32' } });
  expect(quantity.value).toBe('32');
  const link = screen.getByTestId('link-input');
  fireEvent.change(link, { target: { value: 'exampleurl.com/item' } });
  expect(link.value).toBe('exampleurl.com/item');
  const input = screen.getByTestId('description-input');
  fireEvent.change(input, { target: { value: 'this item is great' } });
  expect(input.value).toBe('this item is great');
  fireEvent.click(screen.getByText('Cancel'));
});

test('create post', () => {
  render(<CreatePost />);
  const name = screen.getByTestId('item-name-input');
  fireEvent.change(name, { target: { value: 'Lights' } });
  expect(name.value).toBe('Lights');
  const price = screen.getByTestId('price-input');
  fireEvent.change(price, { target: { value: '34' } });
  expect(price.value).toBe('34');
  const quantity = screen.getByTestId('quantity-input');
  fireEvent.change(quantity, { target: { value: '32' } });
  expect(quantity.value).toBe('32');
  const link = screen.getByTestId('link-input');
  fireEvent.change(link, { target: { value: 'exampleurl.com/item' } });
  expect(link.value).toBe('exampleurl.com/item');
  const input = screen.getByTestId('description-input');
  fireEvent.change(input, { target: { value: 'this item is great' } });
  expect(input.value).toBe('this item is great');
  fireEvent.click(screen.getByText('Post'));
});

test('create post but incomplete fields', () => {
  render(<CreatePost />);
  expect(() => fireEvent.click(screen.getByText('Post')).toThrow('You need to fill in all the blank'));
});

test('snapshot create-post', () => {
  const component = renderer.create(<CreatePost />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
