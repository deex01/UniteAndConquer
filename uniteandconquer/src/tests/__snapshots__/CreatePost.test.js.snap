// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot create-post 1`] = `
<div
  className="create-post-page"
>
  <div
    className="sidebar"
  >
    <div
      className="greeting"
    >
      <i
        className="far fa-user-circle fa-2x"
      />
       
      Hi, Jeremy
    </div>
    <div
      className="profile"
    />
    <div
      className="chat"
    />
    <div
      className="logout"
    />
    <div
      className="ad"
    />
  </div>
  <div
    className="create-post"
  >
    <div
      className="post-input"
    >
      <h3>
        Create a Post
      </h3>
      <div
        className="all-fields"
      >
         
        <div
          className="post-text-fields"
        >
          <div
            className="post-field"
          >
            <div
              className="label"
            >
              Item Name
            </div>
            <input
              data-testid="item-name-input"
              onChange={[Function]}
            />
          </div>
          <div
            className="post-field"
          >
            <div
              className="label"
            >
              Price/Item
            </div>
            <input
              data-testid="price-input"
              onChange={[Function]}
            />
          </div>
          <div
            className="post-field"
          >
            <div
              className="label"
            >
              Target Quantity
            </div>
            <input
              data-testid="quantity-input"
              onChange={[Function]}
            />
          </div>
          <div
            className="post-field"
          >
            <div
              className="label"
            >
              Item Link
            </div>
            <input
              data-testid="link-input"
              onChange={[Function]}
            />
          </div>
        </div>
        <div
          className="description"
        >
          <div
            className="desc-label"
          >
            Description
          </div>
          <textarea
            className="desc"
            data-testid="description-input"
            onChange={[Function]}
          />
        </div>
        <div
          className="create-post-tags"
        >
          <div
            className="tags-label"
          >
            Tags
          </div>
          <div
            className="post-tags"
          >
            <div
              className="tag"
            >
              Tag1
            </div>
            <div
              className="tag"
            >
              Tag2
            </div>
            <div
              className="tag"
            >
              Tag3
            </div>
            <div
              className="tag"
            >
              Tag4
            </div>
            <div
              className="tag"
            >
              Tag5
            </div>
            <div
              className="tag"
            >
              Tag6
            </div>
          </div>
        </div>
      </div>
      <br />
      <div
        className="create-post-bottom"
      >
        <button
          className="create"
          onClick={[Function]}
          type="button"
        >
          Post
        </button>
        <button
          className="cancel"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
